server/src/inmantals/server.py

```diff
@@ -75,22 +216,57 @@ def __init__(self, instream: BaseIOStream, outstream: BaseIOStream, address):
         self.diagnostics_cache: Optional[lsp_types.PublishDiagnosticsParams] = None
         self.supported_symbol_kinds: Optional[Set[lsp_types.SymbolKind]] = None
         # compiler_venv_path is only relevant for versions of core that require a compiler venv. It is ignored otherwise.
-        self.compiler_venv_path: Optional[str] = None
+        self.inmanta_settings: Dict[str, str] = {}
+
+    def add_setting(self, key: str, value: str) -> None:
+        """Register a default value for a setting."""
+        if value is not None:
+            self.inmanta_settings[key] = value
 
-    async def initialize(self, rootPath, rootUri, capabilities: Dict[str, object], **kwargs):  # noqa: N803
+    def get_setting(self, key, folder: Optional[Folder] = None, default: Optional[str] = None) -> Optional[str]:
+        """
+        Fetch a given setting in the provided folder. If the setting isn't found or no folder is provided, this will look
+        for the default value for this setting.
+        """
+        if folder:
+            folder_setting = folder.get_setting(key, default)
+            return folder_setting if folder_setting else self.get_default_setting(key, default)
+        return self.get_default_setting(key, default)
+
+    def get_default_setting(self, key: str, default: Optional[str] = None) -> Optional[str]:
+        """Reads the default value for a setting set by the initializationOptions argument in the initialize method."""
+        logger.info(f"key={key} self.inmanta_settings={self.inmanta_settings}")
+        return self.inmanta_settings.get(key, default)
+
+    def remove_folder(self, folder_uri: str) -> None:
+        logger.info(self._workspace_folders)
+        del self._workspace_folders[folder_uri]
+
+    async def initialize(
+        self, workspaceFolders: Sequence[object], capabilities: Dict[str, object], root_path=None, root_uri=None, **kwargs
+    ):  # noqa: N803
         logger.debug("Init: " + json.dumps(kwargs))
+        logger.info(f"workspaceFolders={workspaceFolders}")
+
+        if root_path:
+            logger.warning("The rootPath parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
+        if root_uri:
+            logger.warning("The rootUri parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
+
+        if workspaceFolders is None:
+            raise InvalidExtensionSetup("No workspace folder specified.")
 
-        if rootPath is None:
-            raise InvalidExtensionSetup("A folder should be opened instead of a file in order to use the inmanta extension.")
+        # Keep track of the folders opened in this workspace
+        self._workspace_folders: Dict[str, Folder] = Folder.unpack_workspaces(workspaceFolders, self)
+
+        if len(self._workspace_folders) > 1:
+            raise NotImplementedError("Working on multiple folders inside a workspace is not supported yet.")
```

Comment:
Do I understand correctly that this PR prepares everything that is needed to implement this but the rest is in the next PR? Some parts of the code seems to be written as if it can already happen that there is more than 1 _workspace_folders. (cleanup, sync_compile_and_anchor_folders). Shouldn't this be part of the next PR to instead of this one?

Reply:
Yes exactly.
I agree this would be more clean to separate this in the next PR

Reply:
ok I don't really care. I just wanted to make sure I wasn't missing something :)